gunicorn/arbiter.py

```diff
@@ -168,31 +171,38 @@ def start(self):
         self.cfg.when_ready(self)
 
     def init_signals(self):
-        """\
-        Initialize master signal handling. Most of the signals
-        are queued. Child signals only wake up the master.
-        """
-        # close old PIPE
-        for p in self.PIPE:
-            os.close(p)
-
-        # initialize the pipe
-        self.PIPE = pair = os.pipe()
-        for p in pair:
-            util.set_non_blocking(p)
-            util.close_on_exec(p)
-
-        self.log.close_on_exec()
-
-        # initialize all signals
         for s in self.SIGNALS:
             signal.signal(s, self.signal)
-        signal.signal(signal.SIGCHLD, self.handle_chld)
 
     def signal(self, sig, frame):
-        if len(self.SIG_QUEUE) < 5:
-            self.SIG_QUEUE.append(sig)
-            self.wakeup()
+        """ Note: Signal handler! No logging allowed. """
+        self.wakeup(due_to_signal=sig)
+
+        # Some UNIXes require SIGCHLD to be reinstalled, see python signal docs
+        if sig == signal.SIGCHLD:
+            signal.signal(sig, self.signal)
+
+    def wakeup(self, due_to_signal=None):
+        """\
+        Wake up the main master loop.
+        """
+        self.SIG_QUEUE.put(due_to_signal or self.WAKEUP_REQUEST)
+
+    def wait_for_signals(self, timeout):
+        """\
+        Returns list of pending signals (or empty list if timeout)
+        """
+        signals = []
+
+        def append_signal(sig):
+            if sig != self.WAKEUP_REQUEST:
+                signals.append(sig)
+        try:
+            append_signal(self.SIG_QUEUE.get(timeout))
```

Comment:
There's a nasty bug here with `timeout` being passed as the `block` arg to `.get()` (`Queue.get(block=True, timeout=None)`) and preventing `.murder_workers()` and `.manage_workers()` from doing the maintenance.

```suggestion
            append_signal(self.SIG_QUEUE.get(timeout=timeout))
```